File: sdk/aip/python/aip_sdk/credentials.py
"""Verifiable Credentials and Presentations."""
import json
from datetime import datetime, timezone
from typing import Dict, Any, List
from .identity import AIPIdentity

def build_credential(issuer: AIPIdentity, holder_did: str, cred_type: str, claims: Dict[str, Any]) -> Dict:
    now = datetime.now(timezone.utc).isoformat()
    body = {"@context": ["https://www.w3.org/ns/credentials/v2"], "type": ["VerifiableCredential", cred_type], "issuer": issuer.did, "validFrom": now, "credentialSubject": {"id": holder_did, **claims}}
    proof_value = issuer.sign(json.dumps(body).encode())
    return {**body, "proof": {"type": "Ed25519Signature2020", "verificationMethod": f"{issuer.did}#keys-1", "proofPurpose": "assertionMethod", "created": now, "proofValue": proof_value}}

def verify_credential(vc: Dict) -> bool:
    proof = vc.get("proof", {})
    body = {k: v for k, v in vc.items() if k != "proof"}
    return AIPIdentity.verify(json.dumps(body).encode(), proof.get("proofValue", ""), vc.get("issuer", ""))

def build_presentation(holder: AIPIdentity, credentials: List[Dict], challenge: str) -> Dict:
    body = {"@context": ["https://www.w3.org/ns/credentials/v2"], "type": ["VerifiablePresentation"], "holder": holder.did, "verifiableCredential": credentials}
    data = (json.dumps(body) + challenge).encode()
    proof_value = holder.sign(data)
    return {**body, "proof": {"type": "Ed25519Signature2020", "verificationMethod": f"{holder.did}#keys-1", "challenge": challenge, "proofValue": proof_value}}

def verify_presentation(vp: Dict, expected_challenge: str) -> bool:
    if vp.get("proof", {}).get("challenge") != expected_challenge: return False
    body = {k: v for k, v in vp.items() if k != "proof"}
    data = (json.dumps(body) + expected_challenge).encode()
    if not AIPIdentity.verify(data, vp["proof"]["proofValue"], vp["holder"]): return False
    return all(verify_credential(vc) for vc in vp.get("verifiableCredential", []))

File: sdk/aip/typescript/src/credentials.ts
/**
 * Verifiable Credentials and Presentations (W3C VC Data Model v2)
 */

import { AIPIdentity } from './identity';

export interface VerifiableCredential {
  '@context': string[];
  type: string[];
  issuer: string;
  validFrom: string;
  credentialSubject: { id: string; [key: string]: any };
  proof: { type: string; verificationMethod: string; proofPurpose: string; created: string; proofValue: string };
}

export interface VerifiablePresentation {
  '@context': string[];
  type: string[];
  holder: string;
  verifiableCredential: VerifiableCredential[];
  proof: { type: string; verificationMethod: string; challenge: string; proofValue: string };
}

export async function buildCredential(params: {
  issuer: AIPIdentity;
  holderDid: string;
  type: string;
  claims: Record<string, any>;
}): Promise<VerifiableCredential> {
  const now = new Date().toISOString();
  const body = {
    '@context': ['https://www.w3.org/ns/credentials/v2'],
    type: ['VerifiableCredential', params.type],
    issuer: params.issuer.did,
    validFrom: now,
    credentialSubject: { id: params.holderDid, ...params.claims },
  };
  const data = new TextEncoder().encode(JSON.stringify(body));
  const proofValue = await params.issuer.sign(data);
  return { ...body, proof: { type: 'Ed25519Signature2020', verificationMethod: `${params.issuer.did}#keys-1`, proofPurpose: 'assertionMethod', created: now, proofValue } };
}

export async function buildPresentation(params: {
  holder: AIPIdentity;
  credentials: VerifiableCredential[];
  challenge: string;
}): Promise<VerifiablePresentation> {
  const body = {
    '@context': ['https://www.w3.org/ns/credentials/v2'],
    type: ['VerifiablePresentation'],
    holder: params.holder.did,
    verifiableCredential: params.credentials,
  };
  const data = new TextEncoder().encode(JSON.stringify(body) + params.challenge);
  const proofValue = await params.holder.sign(data);
  return { ...body, proof: { type: 'Ed25519Signature2020', verificationMethod: `${params.holder.did}#keys-1`, challenge: params.challenge, proofValue } };
}

export async function verifyCredential(vc: VerifiableCredential): Promise<boolean> {
  const { proof, ...body } = vc;
  const data = new TextEncoder().encode(JSON.stringify(body));
  return AIPIdentity.verify(data, proof.proofValue, vc.issuer);
}

export async function verifyPresentation(vp: VerifiablePresentation, expectedChallenge: string): Promise<boolean> {
  if (vp.proof.challenge !== expectedChallenge) return false;
  const { proof, ...body } = vp;
  const data = new TextEncoder().encode(JSON.stringify(body) + expectedChallenge);
  const holderValid = await AIPIdentity.verify(data, proof.proofValue, vp.holder);
  if (!holderValid) return false;
  for (const vc of vp.verifiableCredential) {
    if (!await verifyCredential(vc)) return false;
  }
  return true;
}
